core/src/test/scala/org/apache/spark/deploy/DecommissionWorkerSuite.scala

```diff
@@ -0,0 +1,405 @@
+import org.apache.spark.scheduler.{SparkListener, SparkListenerJobEnd, SparkListenerStageSubmitted, SparkListenerTaskEnd, SparkListenerTaskStart, TaskInfo}
+import org.apache.spark.shuffle.FetchFailedException
+import org.apache.spark.storage.BlockManagerId
+import org.apache.spark.util.Utils
+
+class DecommissionWorkerSuite
+  extends SparkFunSuite
+    with Logging
+    with LocalSparkContext
+    with BeforeAndAfterEach {
+
+  private val conf = new SparkConf()
+  private val securityManager = new SecurityManager(conf)
+
+  private var masterRpcEnv: RpcEnv = null
+  private var master: Master = null
+  private val workerIdToRpcEnvs: mutable.HashMap[String, RpcEnv] = mutable.HashMap.empty
+  private val workers: mutable.ArrayBuffer[Worker] = mutable.ArrayBuffer.empty
+
+  override def beforeEach(): Unit = {
+    super.beforeEach()
+    masterRpcEnv = RpcEnv.create(Master.SYSTEM_NAME, "localhost", 0, conf, securityManager)
+    master = makeMaster()
+  }
+
+  override def afterEach(): Unit = {
+    try {
+      masterRpcEnv.shutdown()
+      workerIdToRpcEnvs.values.foreach(_.shutdown())
+      workerIdToRpcEnvs.clear()
+      master.stop()
+      workers.foreach(_.stop())
+      workers.clear()
+      masterRpcEnv = null
+    } finally {
+      super.afterEach()
+    }
+  }
+
+  test("decommission workers should not result in job failure") {
+    val maxTaskFailures = conf.get(config.TASK_MAX_FAILURES)
+    val numTimesToKillWorkers = maxTaskFailures + 1
+    val numWorkers = numTimesToKillWorkers + 1
+    makeWorkers(numWorkers)
+
+    // Here we will have a single task job and we will keep decommissioning (and killing) the
+    // worker running that task K times. Where K is more than the maxTaskFailures. Since the worker
+    // is notified of the decommissioning, the task failures should be treated as benign and not
+    // the job.
+
+    sc = createSparkContext(appConf)
+    val executorIdToWorkerInfo = getExecutorToWorkerAssignments
+    val taskIdsKilled = new ConcurrentHashMap[Long, Boolean]
+    val listener = new RootStageAwareListener {
+      override def handleRootTaskStart(taskStart: SparkListenerTaskStart): Unit = {
+        val taskInfo = taskStart.taskInfo
+        if (taskInfo.index == 0) {
+          if (taskIdsKilled.size() < numTimesToKillWorkers) {
+            val workerInfo = executorIdToWorkerInfo(taskInfo.executorId)
+            decommissionWorkerOnMaster(workerInfo, "partition 0 must die")
+            killWorkerAfterTimeout(workerInfo, 1)
+            taskIdsKilled.put(taskInfo.taskId, true)
+          }
+        } else {
+          assert(false, s"Unknown task index ${taskInfo.index}")
+        }
+      }
+
+      override def handleRootTaskEnd(taskEnd: SparkListenerTaskEnd): Unit = {
+        val taskInfo = taskEnd.taskInfo
+        if (taskInfo.index == 0) {
+          // If a task has been killed then it shouldn't be successful
+          assert(taskInfo.successful === !taskIdsKilled.get(taskInfo.taskId))
```

Comment:
Oh the reason behind is unboxing: in java the boolean is Boolean and in between this `null` will be silently mapped to `false`, like here [the Long to 0L](http://www.mapdb.org/blog/scala_concurrent_col/?fbclid=IwAR0sRva8WERyLaPpATLCZZSEzm-MafBuyzns9wB7BpqNYBVWE26JSqjeNgs ).

Still please let's not use the `feature` and change the `get` to `getOrDefault` to avoid the confusion when the next reviewer / code reader see this.